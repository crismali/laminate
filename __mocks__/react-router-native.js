/* eslint-env jest */
const { Switch, Route, matchPath } = require.requireActual('react-router-native')
const withRouter = component => component

export { withRouter, Switch, Route, matchPath }
