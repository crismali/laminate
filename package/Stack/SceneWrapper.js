import { Component } from 'react'
import { Animated } from 'react-native'
import PropTypes from 'prop-types'
import { renderChildrenComponentWithContext } from '../renderChildrenComponentWithContext'

export class SceneWrapper extends Component {
  static propTypes = {
    animationValue: PropTypes.instanceOf(Animated.Value).isRequired,
    children: PropTypes.any.isRequired,
    index: PropTypes.number.isRequired
  }

  static childContextTypes = {
    index: PropTypes.number.isRequired,
    interpolateAnimation: PropTypes.func.isRequired,
    setAnimationValue: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      index: this.props.index,
      interpolateAnimation: this.interpolateAnimation,
      setAnimationValue: this.setAnimationValue
    }
  }

  setAnimationValue = (value) => {
    const { animationValue, index } = this.props
    animationValue.setValue(index + value)
  }

  interpolateAnimation = ({ inputRange, outputRange }) => {
    const { animationValue, index } = this.props
    return animationValue.interpolate({
      inputRange: inputRange.map(value => value + index),
      outputRange
    })
  }

  render () {
    return this.props.children
  }
}

export const WithStackAnimation = renderChildrenComponentWithContext('WithStackAnimation', SceneWrapper.childContextTypes)
