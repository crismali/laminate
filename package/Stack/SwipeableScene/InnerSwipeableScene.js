import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { SwipeCatcher } from '../../SwipeCatcher'

export class InnerSwipeableScene extends Component {
  static defaultProps = {
    completeThreshold: 0.5
  }

  static propTypes = {
    children: PropTypes.any.isRequired,
    completeThreshold: PropTypes.number,
    history: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired,
    setAnimationValue: PropTypes.func.isRequired
  }

  handleMove = ({ percentComplete, velocity }) => {
    this.props.setAnimationValue(1 - percentComplete)
  }

  handleGestureRelease = ({ percentComplete, velocity }) => {
    const { completeThreshold, history, location } = this.props

    if (percentComplete > completeThreshold) {
      history.goBack()
    } else {
      history.replace(location)
    }
  }

  handleGestureTerminate = ({ percentComplete, velocity }) => {
    const { history, location } = this.props
    history.replace(location)
  }

  render () {
    return <SwipeCatcher
      {...this.props}
      onMove={this.handleMove}
      onGestureRelease={this.handleGestureRelease}
      onGestureTerminate={this.handleGestureTerminate}
    />
  }
}
