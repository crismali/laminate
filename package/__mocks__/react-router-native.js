/* eslint-env jest */
import React from 'react'

const { Switch, Route, matchPath, StaticRouter } = require.requireActual('react-router-native')
const withRouter = component => component
const MemoryRouter = (props) => <StaticRouter context={{}} {...props} />

export { withRouter, Switch, Route, matchPath, MemoryRouter, StaticRouter }
