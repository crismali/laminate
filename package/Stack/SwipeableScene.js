import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Route } from 'react-router-native'
import { SwipeCatcher } from '../SwipeCatcher'
import { WithStackAnimation } from './SceneWrapper'

class InnerSwipeableScene extends Component {
  static defaultProps = {
    completeThreshold: 0.5
  }

  static propTypes = {
    children: PropTypes.any.isRequired,
    completeThreshold: PropTypes.number,
    history: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired,
    setAnimationValue: PropTypes.func.isRequired
  }

  handleMove = ({ percentComplete, velocity }) => {
    this.props.setAnimationValue(1 - percentComplete)
  }

  handleGestureRelease = ({ percentComplete, velocity }) => {
    const { completeThreshold, history, location } = this.props

    if (percentComplete > completeThreshold) {
      history.goBack()
    } else {
      history.replace(location)
    }
  }

  handleGestureTerminate = ({ percentComplete, velocity }) => {
    const { history, location } = this.props
    history.replace(location)
  }

  render () {
    return <SwipeCatcher
      {...this.props}
      onMove={this.handleMove}
      onGestureRelease={this.handleGestureRelease}
      onGestureTerminate={this.handleGestureTerminate}
    />
  }
}

export class SwipeableScene extends Component {
  render () {
    return <WithStackAnimation>
      {({ setAnimationValue }) => (
        <Route>
          {({ history, location }) => (
            <InnerSwipeableScene
              {...this.props}
              setAnimationValue={setAnimationValue}
              history={history}
              location={location}
            />
          )}
        </Route>
      )}
    </WithStackAnimation>
  }
}
