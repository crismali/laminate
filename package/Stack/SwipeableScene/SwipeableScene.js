import React, { Component } from 'react'
import { Route } from 'react-router-native'
import { WithStackAnimation } from '../SceneWrapper'
import { InnerSwipeableScene } from './InnerSwipeableScene'

export class SwipeableScene extends Component {
  render () {
    return <WithStackAnimation>
      {({ setAnimationValue }) => (
        <Route>
          {({ history, location }) => (
            <InnerSwipeableScene
              {...this.props}
              setAnimationValue={setAnimationValue}
              history={history}
              location={location}
            />
          )}
        </Route>
      )}
    </WithStackAnimation>
  }
}
