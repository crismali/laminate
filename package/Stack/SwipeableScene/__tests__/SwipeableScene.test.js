/* eslint-env jest */
import React, { Component } from 'react'
import { mount } from 'enzyme'
import { StaticRouter } from 'react-router-native'
import { SceneWrapper } from '../../SceneWrapper'
import { SwipeableScene } from '../SwipeableScene'

const Child = () => <div>a child</div>

class ContextMock extends Component {
  static childContextTypes = SceneWrapper.childContextTypes

  getChildContext () {
    return {
      index: 2,
      interpolateAnimation: jest.fn(),
      setAnimationValue: jest.fn()
    }
  }

  render () {
    return this.props.children // eslint-disable-line
  }
}

it('renders properly', () => {
  const subject = mount(<StaticRouter context={{}}>
    <ContextMock>
      <SwipeableScene some='props' are='fun'>
        <Child />
      </SwipeableScene>
    </ContextMock>
  </StaticRouter>)
  expect(subject).toMatchSnapshot()
})
