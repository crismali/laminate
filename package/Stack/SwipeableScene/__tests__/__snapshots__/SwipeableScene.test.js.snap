// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/"
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ContextMock>
      <SwipeableScene
        are="fun"
        some="props"
      >
        <WithStackAnimation>
          <Route>
            <InnerSwipeableScene
              are="fun"
              completeThreshold={0.5}
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                }
              }
              setAnimationValue={[Function]}
              some="props"
            >
              <SwipeCatcher
                are="fun"
                completeThreshold={0.5}
                direction="leftToRight"
                history={
                  Object {
                    "action": "POP",
                    "block": [Function],
                    "createHref": [Function],
                    "go": [Function],
                    "goBack": [Function],
                    "goForward": [Function],
                    "listen": [Function],
                    "location": Object {
                      "hash": "",
                      "pathname": "/",
                      "search": "",
                    },
                    "push": [Function],
                    "replace": [Function],
                  }
                }
                location={
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                  }
                }
                onGestureRelease={[Function]}
                onGestureTerminate={[Function]}
                onMove={[Function]}
                setAnimationValue={[Function]}
                some="props"
              >
                <View
                  onMoveShouldSetPanResponderCapture={[Function]}
                  onPanResponderGrant={[Function]}
                  onPanResponderMove={[Function]}
                  onPanResponderRelease={[Function]}
                  onPanResponderTerminate={[Function]}
                  style={
                    Array [
                      Object {
                        "height": "100%",
                        "position": "absolute",
                        "width": "100%",
                      },
                      undefined,
                    ]
                  }
                >
                  <div>
                    <Child>
                      <div>
                        a child
                      </div>
                    </Child>
                  </div>
                </View>
              </SwipeCatcher>
            </InnerSwipeableScene>
          </Route>
        </WithStackAnimation>
      </SwipeableScene>
    </ContextMock>
  </Router>
</StaticRouter>
`;
